internal/k8s/ingressstatus.go

```diff
@@ -14,45 +14,94 @@
+			WithField("ingress-class", annotation.IngressClass(ing)).
+			Debug("Unmatched ingress class, skip status update")
+		return
+	}
 
 	ing.Status.LoadBalancer = s.Status
 	_, err := s.Client.NetworkingV1beta1().Ingresses(ing.GetNamespace()).UpdateStatus(ing)
 	if err != nil {
 		s.Logger.
 			WithField("name", ing.GetName()).
 			WithField("namespace", ing.GetNamespace()).
-			WithError(err).Error("unable to update status")
+			WithError(err).Error("Unable to update status")
 	}
 }
 
 func (s *IngressStatusUpdater) OnUpdate(oldObj, newObj interface{}) {
-	// Ignoring OnUpdate allows us to avoid the message generated
-	// from the status update.
-
-	// TODO(dfc) handle these cases:
-	// - OnUpdate transitions from an ingress class which is out of scope
-	// to one in scope.
-	// - OnUpdate transitions from an ingress class in scope to one out
-	// of scope.
+
+	oldIng := oldObj.(*v1beta1.Ingress).DeepCopy()
+	newIng := newObj.(*v1beta1.Ingress).DeepCopy()
+
+	// It's important that we only act when we are sure that there has been a change in the
+	// annotation between the old and new versions, otherwise we could generate a status update
+	// in response to another status update.
+	if !annotation.MatchesIngressClass(newIng, s.IngressClass) {
+		// new version of Ingress does not match ingress class, we may need
+		// to clear the status.
+		if annotation.MatchesIngressClass(oldIng, s.IngressClass) {
+			// The old version did match the ingress class, this object has
+			// fallen out of scope. Reset the status info.
+			s.Logger.
+				WithField("name", newIng.GetName()).
+				WithField("namespace", newIng.GetNamespace()).
+				WithField("ingress-class", annotation.IngressClass(newIng)).
+				Debug("Updated Ingress is out of scope, clearing status info")
+			newIng.Status.LoadBalancer = v1.LoadBalancerStatus{}
+			_, err := s.Client.NetworkingV1beta1().Ingresses(newIng.GetNamespace()).UpdateStatus(newIng)
+			if err != nil {
+				s.Logger.
+					WithField("name", newIng.GetName()).
+					WithField("namespace", newIng.GetNamespace()).
+					WithError(err).Error("unable to update status")
+			}
+		}
+		return
+	}
+
+	// New Ingress does match ingress class, check if the old one did as well.
+	if !annotation.MatchesIngressClass(oldIng, s.IngressClass) {
+		// The old version did not match the ingress class, this object has
+		// come into scope. Apply the stored status info.
+		s.Logger.
+			WithField("name", newIng.GetName()).
+			WithField("namespace", newIng.GetNamespace()).
+			WithField("ingress-class", annotation.IngressClass(newIng)).
+			Debug("Updated Ingress is in scope, updating")
+		newIng.Status.LoadBalancer = s.Status
+		_, err := s.Client.NetworkingV1beta1().Ingresses(newIng.GetNamespace()).UpdateStatus(newIng)
+		if err != nil {
+			s.Logger.
+				WithField("name", newIng.GetName()).
+				WithField("namespace", newIng.GetNamespace()).
+				WithError(err).Error("unable to update status")
+		}
+	}
```

Comment:
Given the race condition mentioned above, should we add another condition here that verifies the status is up to date if we own the ingress class? (i.e. to handle the case where someone unkindly removes it)

Reply:
A lot of the complexity here is handled by MatchIngressClass - that handles the case where there's no ingress class specified in the config, which means that Contour will match no annotation, empty annotation, or "contour". If "contour" is present, and is then removed, then it will still match, so this code is not relevant.

If there's a configured ingress class, and someone removes that annotation, then it falls out of our scope, and we should no longer touch it.

tl;dr we only need to handle the "comes into scope" case here, which is neat. Thanks for pointing that out.

Reply:
Yeh I think that makes sense.

I meant, what happens if someone removes the load balancer status that we placed there? I propose that we should put it back, in which case perhaps our policy becomes "update the load balancer if the class of the new object, and the status is stale"? Is there any use case for letting other actor update this?

Reply:
It's actually quite hard to remove the status for Ingress, because it uses the status subresource, which you need to POST to separately. You can't do it just by editing the YAML.

The thing I'm worried about is that we could get an echo effect where we update the status with the loadbalancer, something else updates in response, then we update the loadbalancer again, leading to an update cycle.

Maybe the thing to do is check that the set status matches the one we want to configure, and only update if it doesn't.

Reply:
I've added in a comment explaining the use case and why we left it out.